Cwx.m

```diff
@@ -0,0 +1,360 @@
+
+
+- (bool) setMacro:(int)index macro:(NSString *) msg {
+    if (index < 0 || index > MAX_NUMBER_OF_MACROS - 1) return false;
+    
+    NSString *cmd = [NSString stringWithFormat:@"cwx macro save %i \"%@\"" , index + 1, msg];
+    
+    NSMutableArray * macrosRef = _macros;
+    macrosRef[index] = msg;
+    commandUpdateNotify(cmd, @"macros", _macros, macrosRef);
+    return true;
+}
+
+
+- (bool) getMacro:(int)index macro:(NSString **)string {
+    *string = @"";
+    if (index < 0 || index > MAX_NUMBER_OF_MACROS - 1) return false;
+    *string = (NSString *)_macros[index];
+    return true;
+}
+
+
+- (int) sendMacro:(int) index {
+    if (index < 0 || index > MAX_NUMBER_OF_MACROS - 1) return 0;
+    NSString *cmd = [NSString stringWithFormat:@"cwx macro send %i", index + 1];
+    return [_radio commandToRadio:cmd notify:self];
+}
+
+
+- (void) clearBuffer {
+    [_radio commandToRadio:@"cwx clear"];
+}
+
+
+- (void) erase:(int)numberOfChars {
+    NSString *cmd = [NSString stringWithFormat:@"cwx erase %i", numberOfChars];
+    [_radio commandToRadio:cmd ];
+}
+
+
+- (double) getTxFrequency {
+    @synchronized(_radio.slices) {
+        for ( Slice *slice in _radio.slices) {
+            if (slice.sliceTxEnabled) {
+                return [slice.sliceFrequency doubleValue];
+            }
+        }
+    }
+    return 0.0;
+}
+
+
+- (void) send:(NSString *) string {
+    NSString *cmd = [NSString stringWithFormat:@"cwx send \"%@\"", string];
+    [_radio commandToRadio:cmd];
+}
+
+
+- (void) send:(NSString *) string andBlock:(int) block {
+    NSString *cmd = [NSString stringWithFormat:@"cwx send \"%@\" %i", string, block];
+    [_radio commandToRadio:cmd];
+}
+
+//
+// sendMacroCommand Response
+//     called on the GCD thread associated with the GCD tcpSocketQueue
+//
+//     format: <sequenceNumber>|<errorResponse>|<charPos,block>
+//
+- (void) radioCommandResponse:(uint) seqNum response:(NSString *) response {
+    
+    NSScanner *scan = [[NSScanner alloc] initWithString: response];
+    
+    // get the sequence number... grab it and skip the |
+    NSString *seqNumAsString;
+    [scan scanUpToString:@"|" intoString: &seqNumAsString];
+    [scan scanString:@"|" intoString: nil];
+
+    // get the response error code... grab it and skip the |
+    NSString *errorNumAsString;
+    [scan scanUpToString:@"|" intoString: &errorNumAsString];
+    [scan scanString:@"|" intoString: nil];
+    
+    // Anything other than 0 is an error, just return
+    if ([errorNumAsString intValue] != 0) {
+        return;
+    }
+    
+    // get the rest of the response
+    NSString *remainder;
+    [scan scanUpToString:@"\n" intoString:&remainder];
+    
+    NSArray *values = [remainder componentsSeparatedByString:@","];
+    if ([values count] != 2) return;
+    
+    int charPos = (int)[values[0] integerValue];
+    int block = (int)[values[1] integerValue];
+    
+    // inform the Event Handler (if any)
+    if (_messageQueuedEventDelegate != nil) {
```

Comment:
The delegate handling (here and below) will be invoked on our run queue (or the TCP socket run queue more likely).  Probably worth noting in the Cwx.h file that these will NOT be called on the main dispatch queue.  That way if there are any UI updates that are interested in the update, the user will have to arrange for their own dispatch onto the main queue to make the UI changes there.